### DISCUSSION TITLE
Should anon key work with Supabase Realtime in case my tables do not have RLS enabled?

### DISCUSSION THREAD

**Discussion by TomasHubelbauer** [supabase/supabase#4742] [Questions] [answered]
Hi, I am attempting to get Supabase Realtime working a hobby project to evaluate it.

I have a Supabase project with a database with several tables, all without RLS enabled. I am using the Supabase API using the anon key JWT token. I am attempting to set up Realtime next. I have read through the readme of supabase/realtime and supabase/realtime-js. As per the readmes, I have come up with this code:

```typescript
const url = new URL(process.env.NEXT_PUBLIC_SUPABASE_URL!); // https://{project}.supabase.co
url.protocol = 'wss'; // wss://{project}.supabase.co
url.pathname = 'realtime/v1'; // wss://{project}.supabase.co/realtime/v1
const apikey = process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY!;

const client = new RealtimeClient(url.toString(), { params: { apikey } });
client.connect();
client.onOpen(() => console.log('Socket opened.'));
client.onClose(() => console.log('Socket closed.'));
client.onError((error: Error) => console.log('Socket error', error.message));

const databaseChanges = client.channel('realtime:*');
databaseChanges.on('*', (event: Event) => console.log(event));
databaseChanges.subscribe();
```

This code works insofar that it prints the "Socket opened." message, but it won't print anything when I go into one of my tables and manually add a new row using the Supabase web app.

In one of the supabase/realtime-js issues I have also found a much easier syntax when using @supabase/supabase-js instead of @supabase/realtime-js:

```typescript
const client = createClient(process.env.NEXT_PUBLIC_SUPABASE_URL!, process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY!);
client.from('*').on('*', payload => console.log('payload', payload)).subscribe((status: string, error: Error) => console.log({ status, error }));
```

This code also works to the degree that it will print "SUBSCRIBED" status, but will not actually call the callback when I edit my table's rows manually in the Supabase web app.

My questions:
- Is the anon key the problem here? I can use this key to fetch data from the database using @supabase/supabase-js so I do not think any permission stuff here is going awry and I do not know why it wouldn't be enough for Realtime. Do I need to mint and sign a JWT with the JWT Secret shown in Supabase settings? What fields does this JWT need to have?
  - I also tried using the service role key (which is supposed to be able to bypass RLS) and it resulted in the same outcome - socket opened, but no callbacks fired when I changed the data. This was the case for both snippets.
  - I also tried minting my own JWT using the JWT Secret and this API route which I called in my FE and used its response as the JWT key:
    ```typescript
    import jwt from 'jsonwebtoken';
    import type { NextApiRequest, NextApiResponse } from 'next';
    
    export default async function handler(request: NextApiRequest, response: NextApiResponse) {
      const secret = process.env.SUPABASE_JWT_SECRET!;
      const token = jwt.sign({ sub: "0001", name: "Tomas Hubelbauer", iat: 1516239022, exp: 1718239022 }, secret);
      response.status(200).send(token);
    }
    ```
    This resulted in no connection being even established due to a 401. What do I need to change to make this JWT work in case both anon and service keys won't?
    Note: I also found https://supabase.com/docs/guides/hosting/overview#api-keys which seems to be usable for minting JWT in the UI.
- Do I need to have RLS enabled for Realtime to work at all or can it work without it? Do I need to go and set up Triggers in the Supabase web UI? If so, how and is this documented somewhere?
- What is the difference between using the two code snippets? Is the fact that Realtime seems to be exposed directly through @supabase/supabase-js a nice-to-have feature and is that snippet equivalent to using @supabase/realtime-js or is there a practical difference?

Thank you

**Comment by GaryAustin1** [comment, marked as answer]
As far as anon key goes it is being worked on:
https://github.com/supabase/supabase/discussions/4342

**Reply by TomasHubelbauer** [reply]
The PR linked in that discussion has now been merged, which is great to see. I'll follow up on the release and see if it works for me to enable anon to listed to Realtime. Meanwhile I am still looking for answers to my other questions. I think the service role key did not work for the same reason anon key did - it was not the hard-coded `authenticated` Realtime role. I think RLS needs to be enabled for Realtime to work after this PR has been merged because RLS is how you set the policies, including `auth.role() = 'anon'` for anon key to work? And I am still just as confused about using Realtime through `@supabase/supabase-js` versus `@supabase/realtime-js`. If you happen to know the answers to these, I'd much appreciate it; or if anyone else reading this does, of course!

**Reply by GaryAustin1** [reply]
I'm not up to speed on how realtime will work with anon again.
As far as realtime.js or supabase.js you can go either route, I think supabase.js is a bit clearer and has some helper functions.

**Reply by GaryAustin1** [reply]
@TomasHubelbauer I just saw this and not sure anon is working yet... https://github.com/supabase/supabase/issues/4842

**Reply by TomasHubelbauer** [reply]
It seems as though while the PR for this has been merged already (see supabase/walrus#30), the feature is not rolled out yet. According to the issue you linked I believe anon should work when that happens:

> Soon, Realtime RLS will enable anyone to use any role (e.g. authenticated, anon)

But I am still not sure if it will work with the anon API key or if a JWT will have to be minted. Essentially, if one will be able to use Realtime without Auth. I would like to use Auth, too, but it has a blocker currently which makes it a no-go for me (supabase/gotrue#330) so if Realtime doesn't work with anon without Auth, I'll be toast anyway.

**Reply by w3b6x9** [reply]
@TomasHubelbauer we're currently in the process of deploying the newest Realtime RLS to all Supabase projects. This means that you will be able to use Realtime whether your users are logged in or not. This will work with your existing anon JWTs and there is nothing additional you have to do.

**Reply by TomasHubelbauer** [reply]
Thanks!

**Reply by TomasHubelbauer** [reply]
Hey! It's been around nice months since I created this discussion and I am checking out Supabase Realtime again to see if it will work for me with these changes merged.

I have a basic setup: my Supabase project has a single table, RLS enabled to allow authenticated users, I am using Realtime like so:

```
useEffect(() => {
  const channel = supabase
    .channel("items")
    .on(
      "postgres_changes",
      { event: "*", schema: "public", table: "items" },
      async (payload) => {
        console.log(payload);
      }
    )
    .subscribe();
  return () => {
    channel.unsubscribe();
  };
}, []);
```

The connection succeeds, but no events ever come. I checked to see what token Supabase Realtime is using:

```
    console.log(supabase.realtime.accessToken);
    supabase.auth
      .getSession()
      .then((session) => console.log(session.data.session?.access_token));
```

It is the same key. I can do normal Postgrest (via supabase-js) calls no problem so it should be the token of the authenticated user. I see no tables in the `realtime` schema BTW.

I also tried `realtime-js`. I know it shouldn't make a difference what I use, `supabase-js` or `realtime-js` but I figured if it wasn't working for me with `supabase-js`, `realtime-js` might do the trick:

```
  useEffect(() => {
    void (async function () {
      const { data } = await supabase.auth.getSession();
      const accessToken = data.session?.access_token;
      const client = new RealtimeClient(
        process.env.NEXT_PUBLIC_SUPABASE_REALTIME_URL!,
        {
          params: {
            apikey: accessToken,
            eventsPerSecond: 10,
          },
        }
      );

      const channel = client.channel("db-changes");

      channel.on(
        "postgres_changes",
        { event: "*", schema: "public" },
        (payload) => {
          console.log("All changes in public schema: ", payload);
        }
      );

      channel.subscribe(async (status) => {
        console.log("Channel status: ", status);
        if (status === "SUBSCRIBED") {
          console.log("Ready to receive database changes!");
        }
      });
    })();
```

This never even connects and I get a websocket connection refused error with this. The URL is correct: `NEXT_PUBLIC_SUPABASE_REALTIME_URL=wss://….supabase.co/realtime/v1`.

I tried turning off RLS and it still didn't work.

Are these code snippets something that should work as of today? I am trying to understand whether the problem is with how I am using Supabase Realtime (thought the snippet to use it via `supabase-js` seems pretty straightforward) or if there is a problem with the libraries? I am not sure what else I could try. :/

**Reply by TomasHubelbauer** [reply]
I just realized this discussion was originally about using the anon key which I am no longer doing, I am using Supabase Auth now to get a proper user session. I will create a new discussion with exactly what problem I am facing now.
